Rust/src/lib.rs

```diff
@@ -1,111 +1,147 @@
+    pub fn new(number_of_coconuts: usize, voltage: f32, nailed: bool) -> Self {
+        ParrotConfig {
+            number_of_coconuts: number_of_coconuts,
+            voltage: voltage,
+            nailed: nailed,
         }
     }
 }
 
-fn compute_base_speed_for_voltage(voltage: f32) -> f32 {
-    let fixed_base_speed = 24.0;
-    let base_speed_for_voltage = voltage * base_speed();
-    if base_speed_for_voltage < fixed_base_speed {
-        base_speed_for_voltage
+trait Parrot {
+    fn speed(&self) -> Result<f32, &str>;
+    fn new(conf: ParrotConfig) -> Self;
+}
+
+struct EuropeanParrot {}
+
+impl Parrot for EuropeanParrot {
+    fn speed(&self) -> Result<f32, &str> {
+        Ok(BASE_SPEED)
     }
-    else {
-        fixed_base_speed
+    fn new(_conf: ParrotConfig) -> EuropeanParrot {
+        EuropeanParrot {}
     }
 }
 
-fn load_factor() -> f32 {
-    9.0
+struct AfricanParrot {
+    config: ParrotConfig,
 }
 
-fn base_speed() -> f32 {
-    12.0
+impl Parrot for AfricanParrot {
+    fn speed(&self) -> Result<f32, &str> {
+        Ok(f32::max(
+            BASE_SPEED - LOAD_FACTOR * self.config.number_of_coconuts as f32,
+            0.0,
+        ))
+    }
+    fn new(conf: ParrotConfig) -> AfricanParrot {
+        AfricanParrot { config: conf }
+    }
+}
+
+struct NorwegianBlueParrot {
+    config: ParrotConfig,
+}
+
+impl Parrot for NorwegianBlueParrot {
+    fn speed(&self) -> Result<f32, &str> {
+        match self.config.nailed {
+            true => Ok(0.0),
+            false => Ok(f32::min(self.config.voltage * BASE_SPEED, FIXED_BASE_SPEED)),
+        }
+    }
+    fn new(conf: ParrotConfig) -> NorwegianBlueParrot {
+        NorwegianBlueParrot { config: conf }
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
-    #[test]
-    fn european_parrot_speed() {
-        let parrot = Parrot { parrot_type: "european_parrot",
-                              number_of_coconuts: 0,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 12.0);
+    struct ParrotTest<'a> {
+        expected: f32,
+        config: ParrotConfig,
+        description: &'a str,
     }
 
-    #[test]
-    fn african_parrot_speed_with_one_coconut() {
-        let parrot = Parrot { parrot_type: "african_parrot",
-                              number_of_coconuts: 1,
-                              voltage: 0.0,
-                              nailed: false };
-        assert_eq!(parrot.speed().unwrap(), 3.0);
+    // parrot_test macro to run tests for types that implement the Parrot trait
+    // syntax: ([type that implements Parrot, Vec<ParrotTest>])
+    macro_rules! parrot_test {
+        ($type: ty, $tests:expr) => {{
+            for test in $tests {
+                let parrot = <$type>::new(test.config);
+                println!("{}", test.description);
+                assert_eq!(parrot.speed().unwrap(), test.expected);
+            }
+        }};
     }
```

Comment:
Cool, look so simple here! Regret not sticking with it now, will do this next time.

Does this result in only 1 test being reported by rust/cargo per call to `macro_rules`?

Did you try including the `#[test]` annotation within the macro?

Reply:
Macros seem to be yet another thing in Rust that looks crazy until it suddenly makes sense. This is about as simple as you can get but there is a tonne of stuff you can do.

Very good shout on the test reporting - it currently reports one per call to the macro but i will try this with extra #[test] annotations as you suggest.

Reply:
If it does work it will make macros even cooler

Reply:
> `#[test]` attribute is only allowed on non associated functions

No. Is the answer